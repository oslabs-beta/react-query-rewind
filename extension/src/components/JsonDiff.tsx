import React from 'react';
import JsonFormatter from './JsonFormatter';
import Typography from '@mui/material/Typography';
import Container from '@mui/material/Container';
import jsondiffpatch from 'jsondiffpatch';
import '../css/jsonDiff.css';
import { JsonDiffType } from '../types';


const JsonDiff: React.FC<JsonDiffType> = ({ oldJson, currentJson, queryKey, isHidden }) => {

  // handle scenario where we're on the first state - getting currentJson but not oldJson
  if (currentJson === '') return (
    <Typography
      variant='body1'
      style={{fontStyle: 'italic'}}
    >
      Initial state - no comparison available
    </Typography>
  )

  // get comparison obj
  const delta = jsondiffpatch.diff(oldJson, currentJson);
  // delta is undefined if the 2 objects are the exact same - not sure how I can render this
  // console.log('delta: ', delta);

  if (delta) {
    // Use library's html formatter that generates vanilla CSS
    const htmlFormatter = jsondiffpatch.formatters.html;
    const htmlDiff = htmlFormatter.format(delta, oldJson);
    // React-specific functions to handle raw html
    const createMarkupHtml = () => ({ __html: htmlDiff });

    return (
      <div className={`json-diff-container ${isHidden ? 'jsondiffpatch-unchanged-hidden' : ''}`}>
        <Container>
          <div dangerouslySetInnerHTML={createMarkupHtml()}></div>
        </Container>
      </div>
    )
  }

  // handle errors - this is appearing when there is no change tho?
  return (
    <Typography
      variant='body1'
      style={{fontStyle: 'italic'}}
    >
      QueryKey data not modified on this state change
    </Typography>

  )
}

export default JsonDiff;
