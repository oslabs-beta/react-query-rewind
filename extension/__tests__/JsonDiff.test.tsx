import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import JsonDiff from '../src/components/JsonDiff';

describe('JsonDiff Component', () => {
  it('displays initial state message for empty currentJson and snapshot', () => {
    const { getByText, asFragment } = render(
      <JsonDiff
        oldJson={undefined}
        currentJson=""
        queryKey="testQueryKey"
        isHidden={false}
      />
    );
    expect(getByText(/Initial state - no comparison available/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays a diff when there are differences between oldJson and currentJson and snapshot', () => {
    const oldJson = { key: "oldValue" };
    const currentJson = { key: "newValue" };
    const { container, asFragment } = render(
      <JsonDiff
        oldJson={oldJson}
        currentJson={currentJson}
        queryKey="testQueryKey"
        isHidden={false}
      />
    );
    // Assert that the diff is displayed
    expect(container.querySelector('.jsondiffpatch-delta')).not.toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });
  

  it('displays no changes message when oldJson and currentJson are the same', () => {
    const json = { key: "value" };
    const { getByText, asFragment } = render(
      <JsonDiff
        oldJson={json}
        currentJson={json}
        queryKey="testQueryKey"
        isHidden={false}
      />
    );
    expect(getByText(/QueryKey data not modified on this state change/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('toggles visibility based on isHidden prop', () => {
    const json = { key: "value" };
    const newJson = { key: "newValue" };
    const { container, rerender, asFragment } = render(
      <JsonDiff
        oldJson={json}
        currentJson={newJson}
        queryKey="testQueryKey"
        isHidden={true}
      />
    );
    // Initially, the class indicating hidden changes should be present
    expect(container.querySelector('.jsondiffpatch-unchanged-hidden')).not.toBeNull();
    expect(asFragment()).toMatchSnapshot();
  
    // Rerender with isHidden = false and check if the class is removed
    rerender(
      <JsonDiff
        oldJson={json}
        currentJson={newJson}
        queryKey="testQueryKey"
        isHidden={false}
      />
    );
    expect(container.querySelector('.jsondiffpatch-unchanged-hidden')).toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });
  
});
