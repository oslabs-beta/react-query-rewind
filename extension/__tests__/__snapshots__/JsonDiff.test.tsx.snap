// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonDiff Component displays a diff when there are differences between oldJson and currentJson and snapshot 1`] = `
<DocumentFragment>
  <div
    class="json-diff-container "
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
    >
      <div>
        <div
          class="jsondiffpatch-delta jsondiffpatch-node jsondiffpatch-child-node-type-object"
        >
          <ul
            class="jsondiffpatch-node jsondiffpatch-node-type-object"
          >
            <li
              class="jsondiffpatch-modified"
              data-key="key"
            >
              <div
                class="jsondiffpatch-property-name"
              >
                key
              </div>
              <div
                class="jsondiffpatch-value jsondiffpatch-left-value"
              >
                <pre>
                  "oldValue"
                </pre>
              </div>
              <div
                class="jsondiffpatch-value jsondiffpatch-right-value"
              >
                <pre>
                  "newValue"
                </pre>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`JsonDiff Component displays initial state message for empty currentJson and snapshot 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    style="font-style: italic;"
  >
    Initial state - no comparison available
  </p>
</DocumentFragment>
`;

exports[`JsonDiff Component displays no changes message when oldJson and currentJson are the same 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    style="font-style: italic;"
  >
    QueryKey data not modified on this state change
  </p>
</DocumentFragment>
`;

exports[`JsonDiff Component toggles visibility based on isHidden prop 1`] = `
<DocumentFragment>
  <div
    class="json-diff-container jsondiffpatch-unchanged-hidden"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
    >
      <div>
        <div
          class="jsondiffpatch-delta jsondiffpatch-node jsondiffpatch-child-node-type-object"
        >
          <ul
            class="jsondiffpatch-node jsondiffpatch-node-type-object"
          >
            <li
              class="jsondiffpatch-modified"
              data-key="key"
            >
              <div
                class="jsondiffpatch-property-name"
              >
                key
              </div>
              <div
                class="jsondiffpatch-value jsondiffpatch-left-value"
              >
                <pre>
                  "value"
                </pre>
              </div>
              <div
                class="jsondiffpatch-value jsondiffpatch-right-value"
              >
                <pre>
                  "newValue"
                </pre>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`JsonDiff Component toggles visibility based on isHidden prop 2`] = `
<DocumentFragment>
  <div
    class="json-diff-container "
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
    >
      <div>
        <div
          class="jsondiffpatch-delta jsondiffpatch-node jsondiffpatch-child-node-type-object"
        >
          <ul
            class="jsondiffpatch-node jsondiffpatch-node-type-object"
          >
            <li
              class="jsondiffpatch-modified"
              data-key="key"
            >
              <div
                class="jsondiffpatch-property-name"
              >
                key
              </div>
              <div
                class="jsondiffpatch-value jsondiffpatch-left-value"
              >
                <pre>
                  "value"
                </pre>
              </div>
              <div
                class="jsondiffpatch-value jsondiffpatch-right-value"
              >
                <pre>
                  "newValue"
                </pre>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
